transcode/manifest.go

```diff
@@ -2,46 +2,127 @@ package transcode
+func DownloadRenditionManifest(sourceManifestOSURL string) (m3u8.MediaPlaylist, error) {
+	// Download the manifest
+	rc, err := clients.DownloadOSURL(sourceManifestOSURL)
 	if err != nil {
-		return nil, fmt.Errorf("error decoding manifest: %s", err)
+		return m3u8.MediaPlaylist{}, fmt.Errorf("error downloading manifest: %s", err)
+	}
+
+	// Parse the manifest
+	playlist, playlistType, err := m3u8.DecodeFrom(rc, true)
+	if err != nil {
+		return m3u8.MediaPlaylist{}, fmt.Errorf("error decoding manifest: %s", err)
 	}
 
 	// We shouldn't ever receive Master playlists from the previous section
 	if playlistType != m3u8.MEDIA {
-		return nil, fmt.Errorf("received non-Media manifest, but currently only Media playlists are supported")
+		return m3u8.MediaPlaylist{}, fmt.Errorf("received non-Media manifest, but currently only Media playlists are supported")
 	}
 
 	// The check above means we should be able to cast to the correct type
 	mediaPlaylist, ok := playlist.(*m3u8.MediaPlaylist)
-	if !ok {
-		return nil, fmt.Errorf("failed to parse playlist as MediaPlaylist: %s", err)
+	if !ok || mediaPlaylist == nil {
+		return m3u8.MediaPlaylist{}, fmt.Errorf("failed to parse playlist as MediaPlaylist")
 	}
 
-	// Loop over each segment and convert it from a relative to a full ObjectStore-compatible URL
+	return *mediaPlaylist, nil
+}
+
+// Loop over each segment and convert it from a relative to a full ObjectStore-compatible URL
+func GetSourceSegmentURLs(sourceManifestURL string, manifest m3u8.MediaPlaylist) ([]string, error) {
 	var urls []string
-	for _, segment := range mediaPlaylist.Segments {
+	for _, segment := range manifest.Segments {
 		// The segments list is a ring buffer - see https://github.com/grafov/m3u8/issues/140
 		// and so we only know we've hit the end of the list when we find a nil element
 		if segment == nil {
 			break
 		}
 
-		urls = append(urls, manifestURLToSegmentURL(sourceManifestURL, segment.URI))
+		u, err := manifestURLToSegmentURL(sourceManifestURL, segment.URI)
+		if err != nil {
+			return nil, err
+		}
+		urls = append(urls, u)
 	}
 
 	return urls, nil
 }
 
-func manifestURLToSegmentURL(manifestURL, segmentFilename string) string {
-	i := strings.LastIndex(manifestURL, "/")
-	return manifestURL[:i] + "/" + segmentFilename
+// Generate a Master manifest, plus one Rendition manifest for each Profile we're transcoding, then write them to storage
+func GenerateAndUploadManifests(sourceManifest m3u8.MediaPlaylist, targetManifestOSURL string, transcodeProfiles []cache.EncodedProfile) error {
+	// Generate the base target OS URL to which we can append filenames / subdirectories to
+	baseURL := targetManifestOSURL[:strings.LastIndex(targetManifestOSURL, "/")]
+
+	// Generate the master + rendition output manifests
+	masterPlaylist := m3u8.NewMasterPlaylist()
+
+	for i, profile := range transcodeProfiles {
+		// For each profile, add a new entry to the master manifest
+		masterPlaylist.Append(
```

Comment:
According to the Apple HLS spec, the master playlist is supposed to be in reverse descending order of bitrate then resolution. I think that's how the HLS players need the master playlist to be ordered (?). If so, I don't believe we're guaranteed to get the highest bitrate rendition first from the BOT flow. I added reverse sorting playlist logic in the previous flow that might help if the HLS spec is how it's supposed to be implemented.

Reply:
Good point, I created https://github.com/livepeer/catalyst-api/issues/94 to do it as a follow-up once transcoding is in. I'll have to look at the spec, but I seem to remember that they should be generally ordered like that but that there's some more subtle stuff around which one comes first